compiler/il/OMRSymbolReference.cpp

```diff
@@ -313,7 +313,12 @@ OMR::SymbolReference::hasKnownObjectIndex()
 uintptrj_t*
 OMR::SymbolReference::getKnownObjectReferenceLocation(TR::Compilation *comp)
    {
-   return self()->hasKnownObjectIndex() ?
+#ifdef J9_PROJECT_SPECIFIC
+   bool useKOT = comp->isOutOfProcessCompilation() ? false : true;
+#else
+   bool useKOT = true;
+#endif
+   return (self()->hasKnownObjectIndex() && useKOT) ?
```

Comment:
`getPointerLocation()` is safe to be executed at the server. But where `SymbolReference::getKnownObjectReferenceLocation()` is called might require atomic VM access for multiple operations. The he following code might get executed at the server. 

https://github.com/eclipse/openj9/blob/67fd09e4eceff04511accfd55483ab81d4d7a360/runtime/compiler/optimizer/J9TransformUtil.cpp#L701-L746.

```c++
            baseObjectRefLocation = baseObject->getSymbolReference()->getKnownObjectReferenceLocation(comp);
            }
         else if (baseObject->getOpCode().isLoadVarDirect() && baseObject->getSymbol()->isStatic())
            {
            TR::StaticSymbol *baseSym = baseObject->getSymbol()->castToStaticSymbol();
            if (!baseObject->getSymbolReference()->isUnresolved())
               {
               // Note: There is an implicit assumption here that if we can reach
               // a const object, then its constructor must have completed and its
               // final fields will never change again.  Since an object can't be
               // reached through the constant pool or call site table until its
               // constructor completes, there is no known case where this is false,
               // but if one ever arises, we'll need to check for that case here.
               //
               if (baseSym->isFixedObjectRef())
                  baseObjectRefLocation = (uintptrj_t*)baseSym->getStaticAddress();
               }
            }

         if (!baseObjectRefLocation)
            return NULL; // Nothing left we can do

         //
         // Now we know we have an indirect from an identifiable location.
         // We can use this information to simplify loads of final fields.
         //

         if (isFinalFieldOfNativeStruct(symRef, comp))
            {
            if (baseObjectRefLocation && typeIsConstible)
               {
               if (performTransformation(comp, "O^O transformIndirectLoad: [%p] evaluate native structure walk into const from %s, based on known object node %p\n", node, node->getSymbolReference()->getName(comp->getDebug()), baseObject))
                  {
                  // We have object.a.b.c.d.primitiveField, where a.b.c.d walks through some native data structure.
                  // We can simplify this into a const.
                  //
                  bool loadFailed = false;
                  uint8_t *structure;

                     {
                     // First dereference with VM access to dig a pointer out of an object
                     TR::VMAccessCriticalSection digPointerOutOfObject(comp->fej9());

                     uintptrj_t baseObjectRef = baseObject->getSymbol()->isStatic() ?
                                                comp->fej9()->getStaticReferenceFieldAtAddress((uintptrj_t)baseObjectRefLocation) :
                                                *baseObjectRefLocation;
```

Reply:
I don't see it from the code you pasted. SymbolReference::getKnownObjectReferenceLocation() is not under a VMAccess critical section, unless I missed it

Reply:
However, below that piece of code there is some nasty sequence that is not treated properly at the server

Reply:
`SymbolReference::getKnownObjectReferenceLocation()` is not under a VMAccess. But if `baseObjectRefLocation` is set by `SymbolReference::getKnownObjectReferenceLocation()`, there is a chance the following code will get executed at the server. On a second thought, maybe I should not change `SymbolReference::getKnownObjectReferenceLocation()` but add the check on isOutOfProcessCompilation in the following code. 

```c++
            if (baseObjectRefLocation && typeIsConstible)
               {
               if (performTransformation(comp, "O^O transformIndirectLoad: [%p] evaluate native structure walk into const from %s, based on known object node %p\n", node, node->getSymbolReference()->getName(comp->getDebug()), baseObject))
                  {
                  // We have object.a.b.c.d.primitiveField, where a.b.c.d walks through some native data structure.
                  // We can simplify this into a const.
                  //
                  bool loadFailed = false;
                  uint8_t *structure;

                     {
                     // First dereference with VM access to dig a pointer out of an object
                     TR::VMAccessCriticalSection digPointerOutOfObject(comp->fej9());

                     uintptrj_t baseObjectRef = baseObject->getSymbol()->isStatic() ?
                                                comp->fej9()->getStaticReferenceFieldAtAddress((uintptrj_t)baseObjectRefLocation) :
                                                *baseObjectRefLocation;
```

Reply:
agreed